### PR TITLE
feat!: Support Decimal types

### PR DESCRIPTION
**Pull request by disq**
Maybe a discussion is needed about keeping/removing the Uint64 special case.

### CODE REVIEW

File: plugins/destination/postgresql/pgarrow/to_arrow.go [hunk 1]
```diff
@@ -240,3 +234,41 @@ func parseTime(t string) (arrow.DataType, bool) {
 		return arrow.FixedWidthTypes.Time64us, true
 	}
 }
+
+func parseNumeric(t string) (arrow.DataType, bool) {
```
Comment (erezrokah) [inline plugins/destination/postgresql/pgarrow/to_arrow.go:238]: MySQL does something similar https://github.com/cloudquery/cloudquery/blob/6dba5b42f2058aed2219f95b0aff876aff1ab6b0/plugins/destination/mysql/client/types.go#L26

File: plugins/destination/postgresql/pgarrow/to_arrow.go [hunk 2]
```diff
@@ -240,3 +234,41 @@ func parseTime(t string) (arrow.DataType, bool) {
...
+		panic("precision cannot be 0")
+	}
+	scale, err := strconv.ParseInt(matches[0][2], 10, 64)
+	if err != nil {
+		panic("error parsing scale " + err.Error())
+	}
+
+	if precision == 20 && scale == 0 {
+		// special case
+		return arrow.PrimitiveTypes.Uint64, true
```
Comment (disq) [inline plugins/destination/postgresql/pgarrow/to_arrow.go:261]: Keeping the `numeric(20,0) => Uint64` special case here.

File: plugins/destination/postgresql/pgarrow/to_arrow.go [hunk 3]
```diff
@@ -240,3 +234,41 @@ func parseTime(t string) (arrow.DataType, bool) {
...
+	if precision <= 38 {
+		d, _ := arrow.NewDecimalType(arrow.DECIMAL128, int32(precision), int32(scale))
+		return d, true
+	}
+	if precision <= 76 {
+		d, _ := arrow.NewDecimalType(arrow.DECIMAL256, int32(precision), int32(scale))
+		return d, true
+	}
+
+	return arrow.BinaryTypes.String, true
```
Comment (disq) [inline plugins/destination/postgresql/pgarrow/to_arrow.go:273]: Can't represent these in Arrow so we fall back to string

File: plugins/destination/postgresql/pgarrow/to_arrow_test.go [hunk 4]
```diff
@@ -93,12 +93,14 @@ func TestPg10ToArrow(t *testing.T) {
 		{"time(6) with time zone", arrow.FixedWidthTypes.Time64us},
 		// special case for uint64
 		{"numeric(20,0)", arrow.PrimitiveTypes.Uint64},
+		{"numeric", mustDec(arrow.NewDecimalType(arrow.DECIMAL128, 38, 0))},
+		{"numeric (1, 0)", mustDec(arrow.NewDecimalType(arrow.DECIMAL128, 1, 0))},
+		{"numeric(38, 15)", mustDec(arrow.NewDecimalType(arrow.DECIMAL128, 38, 15))},
+		{"numeric (50, 25)", mustDec(arrow.NewDecimalType(arrow.DECIMAL256, 50, 25))},
```
Comment (disq) [inline plugins/destination/postgresql/pgarrow/to_arrow_test.go:99]: Testing different spacing here, both `numeric(` and `numeric (`

**Review by hermanschaaf — APPROVED** [review]

**Comment by hermanschaaf** [comment]
Should we mark this as a breaking change? Because it will change the column types from text to numeric

File: plugins/destination/postgresql/pgarrow/to_arrow.go [hunk 5]
```diff
@@ -240,3 +234,37 @@ func parseTime(t string) (arrow.DataType, bool) {
...
+	if err != nil {
+		panic("error parsing scale " + err.Error())
+	}
+
+	if precision == 20 && scale == 0 {
+		// special case
+		return arrow.PrimitiveTypes.Uint64, true
+	}
+
+	if precision <= 76 {
```
Comment (candiduslynx) [inline plugins/destination/postgresql/pgarrow/to_arrow.go:264]: If the prec is<=38 you can use decimal128, btw

Comment (disq) [reply plugins/destination/postgresql/pgarrow/to_arrow.go:264]: Removed in https://github.com/cloudquery/cloudquery/pull/17247/commits/65118a48342d563dbc9ca365b5a347b0876a4cb9 to simplify

File: plugins/destination/postgresql/pgarrow/to_arrow.go [hunk 6]
```diff
@@ -240,3 +234,37 @@ func parseTime(t string) (arrow.DataType, bool) {
...
+func parseNumeric(t string) (arrow.DataType, bool) {
+	matches := reNumeric.FindAllStringSubmatch(t, -1)
+	if len(matches) == 0 {
+		return nil, false
+	}
+
+	if len(matches[0]) < 3 || matches[0][1] == "" {
+		// no precision/scale specified
+		d, _ := arrow.NewDecimalType(arrow.DECIMAL256, 38, 0)
+		return d, true
```
Comment (candiduslynx) [inline plugins/destination/postgresql/pgarrow/to_arrow.go:247]:
```suggestion
		return &arrow.Decimal256Type{Precision: 38, Scale: 0}, true
```

File: plugins/destination/postgresql/pgarrow/to_arrow.go [hunk 7]
```diff
@@ -240,3 +234,37 @@ func parseTime(t string) (arrow.DataType, bool) {
...
+	}
+
+	if precision == 20 && scale == 0 {
+		// special case
+		return arrow.PrimitiveTypes.Uint64, true
+	}
+
+	if precision <= 76 {
+		d, _ := arrow.NewDecimalType(arrow.DECIMAL256, int32(precision), int32(scale))
+		return d, true
```
Comment (candiduslynx) [inline plugins/destination/postgresql/pgarrow/to_arrow.go:266]:
```suggestion
		return &arrow.Decimal256Type{Precision: int32(precision), Scale: int32(scale)}, true
```

File: plugins/destination/postgresql/pgarrow/to_arrow_test.go [hunk 8]
```diff
@@ -223,16 +225,18 @@ func TestCockroachToArrow(t *testing.T) {
 		{"time(6) with time zone", arrow.FixedWidthTypes.Time64us},
 		// special case for uint64
 		{"numeric(20,0)", arrow.PrimitiveTypes.Uint64},
+		{"numeric", mustDec(arrow.NewDecimalType(arrow.DECIMAL256, 38, 0))},
```
Comment (candiduslynx) [inline plugins/destination/postgresql/pgarrow/to_arrow_test.go:228]: please use `arrow.Decimal256Type` directly

File: plugins/destination/postgresql/pgarrow/to_pg_test.go [hunk 9]
```diff
@@ -40,6 +40,10 @@ func TestArrowToPg10(t *testing.T) {
 		{"time without time zone[]", arrow.ListOf(arrow.FixedWidthTypes.Time64ns)},
 		// special case for uint64
 		{"numeric(20,0)", arrow.PrimitiveTypes.Uint64},
+		{"numeric(38,0)", mustDec(arrow.NewDecimalType(arrow.DECIMAL256, 38, 0))},
```
Comment (candiduslynx) [inline plugins/destination/postgresql/pgarrow/to_pg_test.go:43]: use `arrow.Decimal256Type` directly

**Review by candiduslynx — CHANGES_REQUESTED** [review]

**Review by candiduslynx — APPROVED** [review]

**Pull request merged**